dwave/optimization/src/nodes/linear_algebra.cpp

```diff
@@ -0,0 +1,339 @@
+// Copyright 2025 D-Wave Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include "dwave-optimization/nodes/linear_algebra.hpp"
+
+#include <algorithm>
+#include <array>
+#include <ranges>
+
+#include "../functional_.hpp"
+#include "_state.hpp"
+#include "dwave-optimization/array.hpp"
+#include "dwave-optimization/state.hpp"
+
+namespace dwave::optimization {
+
+////////////////////// MatrixMultiplyNode
+
+// Valid shapes to multiply, and the resulting shape
+// (-1, 2, 5, 3) and (-1, 2, 3, 7) -> (-1, 2, 5, 7)
+// (-1, 3) and (3) -> (-1)
+// (-1, 3) and (3, 7) -> (-1, 7)
+// (-1) and (-1) -> ()
+// (-1) and (-1, 5) -> (5)
+
+ssize_t size_from_shape(std::span<const ssize_t> shape) {
```

Comment:
https://github.com/dwavesystems/dwave-optimization/blob/d1b5555abacbb7ba5fad159261a8196fdd95dd5f/dwave/optimization/include/dwave-optimization/array.hpp#L488
I also keep forgetting it exists.

Though I think maybe you're using the fact that it will return a negative number in the case of dynamic arrays?